### DISCUSSION TITLE
Power: Device PSU Allocation does not change allocation of parent PDU.

### DISCUSSION THREAD

**Discussion by MCDELTAT** [netbox-community/netbox#13840] [Q&A]
Hi all. I was debating filing this as a bug but not sure it was ever the intended behavior affter looking at the code.

I have the following Power Chain:
1. Power Panels called UPS A, UPS B, UPS C
2. Power Feeds: BM49 F1 connected to Power Panel UPS C, BM 49 F2 connected to Power Panel UPS A
3. (2) Non-Racked Devices: 2 APC APF8966D PDU Devices called "PDU BM49 Left" and "PDU BM49 Right"
     a. Input (Power Ports): 1 3P+E 9H cable. Maximum Power Draw 9984W (208V @ 48A), Allocation of 0W.
     b. Output (Power Outlets): 20 C13, 12 C19
4. Dell R650 Device: 2 PSUs. 1 Connected w/ Allocation of 330W, 1 Disconnected.

They are all cabled correctly (dashes indicate a component of the device/object).
Device: Dell R650 - PSU1 (Alloc 330W) -> Device: PDU BM49 Left - B1/1 (Alloc 0W) -> Feed: BFM 49 F1 -> Panel: UPS C

On the PDU Device "PDU BM49 Left" under the Power Utilization block, it says "Input: Input 1, Allocated: 0VA, Available: 14410VA"

Am I doing this right? Seems like the code only cares about input (Power Port Allocations, which consider the PDU as a whole). But I'd much prefer that it be a sum of all the Power Outlets, that way if I change the location of a server or device, it automatically updates and I don't have to do the math in my head. "Was 6000W, moved two 330W devices, so manually update the PDU to be allocated 5340W now."

This also affects things at the rack level. The rack currently reports 0.0%, whereas if Netbox were looking at devices, it would be 2.75% (330W / 1200W)

**Comment by candlerb** [comment]
The power calculation functionality in Netbox is not well documented. I had a go at reverse engineering it here: #12837. IIRC, rack level calculations only take into account power feeds assigned to that rack.

**Reply by candlerb** [reply]
I've just uploaded a custom script [here](https://github.com/netbox-community/customizations/pull/96/files) which might be helpful, or at least a starting point, if your main concern is calculating total power utilisation for a site.

**Reply by MCDELTAT** [reply]
Nice. Thanks for the link will have to read through your reverse engineering tests, but read the conversation for now. I figured that was the opinion of the maintainers given how it seemed to work. That it's easier to go to the PDU and get the per PDU power draw and if you need to update, it's fewer things to hunt down. 

I got my value at the device level from my iDRAC (Dell OOB), which keeps up to a week of data. The average across the day and a week were both 218W, so seemed stable with exception of power on peaks. The peak was a day I know I restarted, so it made sense to accept the average. I had already written a Python script to grab that data and toss it in a CSV so I could evaluate quickly over time.
